code/game/objects/items/weapons/thurible.dm

```diff
@@ -0,0 +1,173 @@
+	. = ..()
+	light()
+
+/obj/item/thurible/on_reagent_change()
+	. = ..()
+	if(!corrupted)
+		var/found_forbidden_reagent = FALSE
+		for(var/datum/reagent/R as anything in reagents.reagent_list)
+			if(R.id == "unholywater")
+				corrupted = TRUE
+				visible_message(
+					"<span class='notice'>You corrupt [src] with unholy water!</span>",
+					"<span class='warning'>You hear a strange gurgling.</span>"
+				)
+				return
+			if(!safe_chem_list.Find(R.id))
+				reagents.del_reagent(R.id)
+				found_forbidden_reagent = TRUE
+		if(found_forbidden_reagent)
+			visible_message(
+					"<span class='notice'>[src] banishes an unholy substance!</span>",
+					"<span class='warning'>You hear a strange fizzing.</span>"
+				)
+
+/// Lights the thurible and starts processing reagents
+/obj/item/thurible/proc/light(mob/user)
+	if(lit)
+		to_chat(user, "<span class='warning'>[src] is already lit!</span>")
+		return
+
+	if(!reagents.total_volume)
+		to_chat(user, "<span class='warning'>[src] is out of fuel!</span>")
+		return
+
+	// Plasma explodes when exposed to fire.
+	if(reagents.get_reagent_amount("plasma"))
+		var/datum/effect_system/reagents_explosion/E = new()
+		E.set_up(round(reagents.get_reagent_amount("plasma") / 2.5, 1), get_turf(src), 0, 0)
+		E.start()
+		return
+
+	// Fuel explodes, too, but much less violently.
+	if(reagents.get_reagent_amount("fuel"))
+		var/datum/effect_system/reagents_explosion/E = new()
+		E.set_up(round(reagents.get_reagent_amount("fuel") / 5, 1), get_turf(src), 0, 0)
+		E.start()
+		return
+
+	// And black powder... but more violently.
+	if(reagents.get_reagent_amount("blackpowder"))
+		var/datum/effect_system/reagents_explosion/E = new()
+		E.set_up(round(reagents.get_reagent_amount("blackpowder") / 2, 1), get_turf(src), 0, 0)
+		E.start()
+		return
+
+	lit = TRUE
+	reagents.set_reacting(TRUE)
+	reagents.handle_reactions()
+	START_PROCESSING(SSobj, src)
+	set_light(2, 0.3, "#E38F46")
+	update_appearance()
+	return TRUE
+
+/// Extinguishes the thurible and stops processing
+/obj/item/thurible/proc/put_out(mob/user)
+	lit = FALSE
+	STOP_PROCESSING(SSobj, src)
+	set_light(0)
+	update_appearance()
+	return TRUE
+
+/// Spreads reagents in a 3x3 area centered on the thurible
+/obj/item/thurible/proc/swing()
+	var/obj/released_reagents = new
+	released_reagents.create_reagents(2)
+	reagents.trans_to(released_reagents, swing_reagents_consumed)
+	var/list/mobs_to_smoke = list()
+	for(var/atom/A in view(1, get_turf(src)))
+		released_reagents.reagents.reaction(A)
+		if(iscarbon(A))
+			var/mob/living/carbon/C = A
+			if(C.can_breathe_gas())
+				mobs_to_smoke += C
+
+	var/percentage_to_add = released_reagents.reagents.total_volume / length(mobs_to_smoke) // Divide the amount of reagents spread around by the number of people inhaling it
+
+	for(var/mob/living/carbon/smoker as anything in mobs_to_smoke)
+		released_reagents.reagents.copy_to(smoker, percentage_to_add)
+
+	if(!reagents.total_volume)
```

Comment:
Probably better to check if it's `<= 0`, it's visually better because it explains that `total_volume` is a number